src/scripts/components/header/SearchAutocomplete.js

```diff
@@ -0,0 +1,127 @@
+import React from 'react';
+import classNames from 'classnames';
+import SurahsStore from 'stores/SurahsStore';
+import connectToStores from 'fluxible-addons-react/connectToStores';
+import request from 'superagent';
+import Settings from 'constants/Settings';
+
+class SearchAutocomplete extends React.Component {
+  constructor(props, context) {
+    super(props, context);
+    this.state = { value: '', ayat: [], surahs: [] };
+    this.cached = {};
+  }
+  componentWillReceiveProps( nextProps ) {
+    //console.log('searchautocomplete componentWillReceiveProps', nextProps);
+    this.setState({ value: nextProps.value });
+    this.suggest( nextProps.value );
+  };
+
+  suggest(value) {
+    if (value.length >= 2) {
+      this.handleSurahSuggestions(value);
+      if (value.length >= 3) {
+        this.handleAyahSuggestions(value);
+      }
+      else {
+        this.setState({ ayat: [] });
+      }
+    }
+    else {
+      this.setState( { surahs: [] } );
```

Comment:
not a fan of setting state every key down, even if not needed. It should set state if length less than 3 & `this.state.surahs.length`